File: examples/typescript/quickstart.ts
/**
 * Next Immo Lab API v1 — TypeScript Quickstart
 * ==============================================
 *
 * Full example: trigger a call, poll for results, webhook handler.
 * Zero dependencies — uses native fetch (Node 18+).
 *
 * Usage:
 *   export NEXTIMMOLAB_API_KEY="ak_live_xxxx"
 *   npx tsx quickstart.ts
 */

const API_BASE = "https://immo.next-lab.tech/api/v1";
const API_KEY = process.env.NEXTIMMOLAB_API_KEY ?? "";

if (!API_KEY) {
  console.error("Set NEXTIMMOLAB_API_KEY environment variable");
  process.exit(1);
}

const headers = {
  "x-api-key": API_KEY,
  "Content-Type": "application/json",
};

// ─── Types ──────────────────────────────────────────────────────────────────

interface TriggerResponse {
  callId: string;
  status: string;
  externalRef?: string;
  roomName?: string;
}

interface CriterionResult {
  criterionId: string;
  question: string;
  answer: string;
  passed: boolean;
  eliminatory: boolean;
}

interface TranscriptEntry {
  role: "agent" | "user";
  text: string;
  timestamp: string;
}

interface CallResult {
  id: string;
  status: "pending" | "ringing" | "in_progress" | "completed" | "failed";
  externalRef?: string;
  leadName: string;
  leadPhone: string;
  leadEmail: string;
  startedAt: string;
  endedAt?: string;
  duration?: string;
  score?: number;
  qualified?: boolean;
  criteriaResults: CriterionResult[];
  transcript?: TranscriptEntry[];
}

interface UsageStats {
  totalCalls: number;
  maxCalls: number | null;
  remaining: number | null;
  byStatus: Record<string, number>;
  activeCalls: unknown[];
  averageScore: number | null;
  qualifiedCount: number;
}

// ─── 1. Check usage ────────────────────────────────────────────────────────

async function checkUsage(): Promise<UsageStats> {
  const res = await fetch(`${API_BASE}/usage`, { headers });
  if (!res.ok) throw new Error(`Usage check failed: ${res.status}`);
  const usage: UsageStats = await res.json();

  console.log(`Calls: ${usage.totalCalls}/${usage.maxCalls ?? "unlimited"}`);
  console.log(`Remaining: ${usage.remaining ?? "unlimited"}`);
  console.log(`Qualified: ${usage.qualifiedCount} | Avg score: ${usage.averageScore ?? "--"}`);

  if (usage.remaining !== null && usage.remaining <= 0) {
    throw new Error("No calls remaining");
  }
  return usage;
}

// ─── 2. Trigger a call ─────────────────────────────────────────────────────

async function triggerCall(): Promise<string> {
  const tomorrow = new Date(Date.now() + 86400000).toISOString().split("T")[0];
  const dayAfter = new Date(Date.now() + 172800000).toISOString().split("T")[0];

  const payload = {
    lead: {
      prenom: "Lukas",
      nom: "Weber",
      telephone: "+41791234567",
      email: "lukas.weber@example.com",
      language: "de" as const,
      budget: "3000",
      remarks: "Sucht hundefreundliche Wohnung, arbeitet von zu Hause",
    },
    property: {
      title: "4.5 Zimmer Wohnung in Zürich",
      address: "Bahnhofstrasse 50, 8001 Zürich",
      price: 2800,
      type: "rent" as const,
      description:
        "Moderne 4.5-Zimmer-Wohnung, 95m2, offene Küche, 2 Schlafzimmer, " +
        "Büro, grosser Balkon mit Blick auf die Limmat. " +
        "Nahe Hauptbahnhof, Einkaufsmöglichkeiten. Keller und Waschküche.",
      features: ["Balkon", "Offene Küche", "Tiefgarage", "Keller", "Haustiere erlaubt"],
    },
    agent_config: {
      agent_name: "Anna",
      agency_name: "Avendo Immobilier",
      transfer_number: "+41791234568",
      qualification_criteria: [
        {
          question: "Was ist Ihr maximales monatliches Budget für die Miete?",
          type: "number",
          expectedValue: "2800",
          eliminatory: true,
        },
        {
          question: "Haben Sie einen festen Arbeitsvertrag oder ein regelmässiges Einkommen?",
          type: "yes_no",
          expectedValue: "oui",
          eliminatory: true,
        },
        {
          question: "Wie viele Personen werden in der Wohnung wohnen?",
          type: "number",
          expectedValue: "1-4",
          eliminatory: false,
        },
      ],
    },
    viewing_slots: [
      { date: tomorrow, startTime: "14:00", endTime: "15:00", maxBookings: 3 },
      { date: dayAfter, startTime: "10:00", endTime: "11:00", maxBookings: 3 },
    ],
    callback_webhook: "https://your-server.com/webhooks/nextimmolab",
    external_ref: `TS-${Date.now()}`,
  };

  console.log(`\nTriggering call to ${payload.lead.prenom} ${payload.lead.nom}...`);

  const res = await fetch(`${API_BASE}/calls/trigger`, {
    method: "POST",
    headers,
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    const err = await res.text();
    throw new Error(`Trigger failed (${res.status}): ${err}`);
  }

  const data: TriggerResponse = await res.json();
  console.log(`Call triggered — ID: ${data.callId} | Status: ${data.status}`);
  return data.callId;
}

// ─── 3. Poll for results ───────────────────────────────────────────────────

async function waitForResult(callId: string, timeoutMin = 10): Promise<CallResult | null> {
  console.log(`\nPolling call ${callId.slice(0, 8)}...`);
  const deadline = Date.now() + timeoutMin * 60_000;

  while (Date.now() < deadline) {
    const res = await fetch(`${API_BASE}/calls/${callId}`, { headers });
    if (!res.ok) throw new Error(`Poll failed: ${res.status}`);

    const call: CallResult = await res.json();
    const time = new Date().toLocaleTimeString();
    process.stdout.write(`  [${time}] ${call.status}`);

    if (["completed", "failed", "no_answer"].includes(call.status)) {
      console.log();
      return call;
    }

    console.log(" — waiting 10s...");
    await new Promise((r) => setTimeout(r, 10_000));
  }

  console.log("Timeout reached.");
  return null;
}

// ─── 4. Display results ────────────────────────────────────────────────────

function printResults(call: CallResult) {
  console.log("\n" + "=".repeat(60));
  console.log(`RESULT: ${call.status.toUpperCase()}`);
  console.log("=".repeat(60));
  console.log(`Lead:      ${call.leadName}`);
  console.log(`Duration:  ${call.duration ?? "--"}`);
  console.log(`Score:     ${call.score ?? "--"}/100`);
  console.log(`Qualified: ${call.qualified ?? "--"}`);

  if (call.criteriaResults.length) {
    const passed = call.criteriaResults.filter((c) => c.passed).length;
    console.log(`\nCriteria (${passed}/${call.criteriaResults.length} passed):`);
    for (const c of call.criteriaResults) {
      const icon = c.passed ? "PASS" : "FAIL";
      const elim = c.eliminatory ? " [ELIMINATORY]" : "";
      console.log(`  [${icon}] ${c.question}${elim}`);
      console.log(`         Answer: ${c.answer ?? "--"}`);
    }
  }

  if (call.transcript?.length) {
    console.log(`\nTranscript (${call.transcript.length} messages):`);
    for (const t of call.transcript.slice(0, 6)) {
      const role = t.role === "agent" ? "AGENT" : "LEAD ";
      const text = t.text.length > 80 ? t.text.slice(0, 80) + "..." : t.text;
      console.log(`  ${role}: ${text}`);
    }
    if (call.transcript.length > 6) {
      console.log(`  ... +${call.transcript.length - 6} more messages`);
    }
  }
  console.log("=".repeat(60));
}

// ─── 5. List recent calls ──────────────────────────────────────────────────

async function listCalls(limit = 5) {
  const res = await fetch(`${API_BASE}/calls?limit=${limit}`, { headers });
  if (!res.ok) throw new Error(`List failed: ${res.status}`);

  const data = await res.json();
  console.log(`\nRecent calls (${data.total} total):`);
  for (const c of data.calls) {
    const score = c.status === "completed" ? `score=${c.score ?? "--"}` : "";
    console.log(
      `  ${c.id.slice(0, 8)}... | ${c.status.padEnd(12)} | ${c.leadName.padEnd(20)} | ${score}`
    );
  }
}

// ─── Webhook handler example (Express) ─────────────────────────────────────

/*
import express from "express";
const app = express();
app.use(express.json());

app.post("/webhooks/nextimmolab", (req, res) => {
  const { event, callId, externalRef, score, qualified, criteriaResults } = req.body;

  console.log(`Webhook: ${event} | Call: ${callId} | Ref: ${externalRef}`);
  console.log(`Score: ${score} | Qualified: ${qualified}`);

  // Sync back to your CRM here
  // e.g. updateLeadInCRM(externalRef, { score, qualified, criteriaResults });

  res.sendStatus(200);
});

app.listen(3001);
*/

// ─── Main ───────────────────────────────────────────────────────────────────

async function main() {
  await checkUsage();
  const callId = await triggerCall();

  console.log("\nWaiting 5s for call to connect...");
  await new Promise((r) => setTimeout(r, 5000));

  const result = await waitForResult(callId);
  if (result) printResults(result);

  await listCalls();
}

main().catch(console.error);

File: examples/python/quickstart.py
#!/usr/bin/env python3
"""
Next Immo Lab API v1 — Python Quickstart
=========================================

Full example: trigger a call, poll for results, handle webhook.

Requirements:
    pip install requests

Usage:
    export NEXTIMMOLAB_API_KEY="ak_live_xxxx"
    python quickstart.py
"""

import os
import sys
import json
import time
import requests
from datetime import datetime, timedelta

# ─── Config ──────────────────────────────────────────────────────────────────

API_BASE = "https://immo.next-lab.tech/api/v1"
API_KEY = os.environ.get("NEXTIMMOLAB_API_KEY", "")

if not API_KEY:
    print("Set NEXTIMMOLAB_API_KEY environment variable")
    sys.exit(1)

HEADERS = {
    "x-api-key": API_KEY,
    "Content-Type": "application/json",
}


# ─── 1. Check usage before triggering ───────────────────────────────────────

def check_usage():
    """GET /api/v1/usage — verify you have remaining calls."""
    res = requests.get(f"{API_BASE}/usage", headers=HEADERS)
    res.raise_for_status()
    usage = res.json()

    print(f"Calls used: {usage['totalCalls']}/{usage.get('maxCalls', 'unlimited')}")
    print(f"Remaining:  {usage.get('remaining', 'unlimited')}")
    print(f"Qualified:  {usage['qualifiedCount']} | Avg score: {usage.get('averageScore', '--')}")

    if usage.get("remaining") is not None and usage["remaining"] <= 0:
        print("No calls remaining — contact support.")
        sys.exit(1)

    return usage


# ─── 2. Trigger a call ──────────────────────────────────────────────────────

def trigger_call():
    """POST /api/v1/calls/trigger — full example with all fields."""

    tomorrow = (datetime.now() + timedelta(days=1)).strftime("%Y-%m-%d")
    day_after = (datetime.now() + timedelta(days=2)).strftime("%Y-%m-%d")

    payload = {
        "lead": {
            "prenom": "Marie",
            "nom": "Müller",
            "telephone": "+41791234567",
            "email": "marie.muller@example.com",
            "language": "fr",
            "budget": "2800",
            "remarks": "Recherche appartement pet-friendly, travaille a domicile",
        },
        "property": {
            "title": "Appartement 3.5 pieces a Lausanne",
            "address": "Avenue de la Gare 22, 1003 Lausanne",
            "price": 2500,
            "type": "rent",
            "description": (
                "Bel appartement renove de 78m2, 2 chambres, salon lumineux, "
                "cuisine equipee, balcon avec vue sur les Alpes. "
                "Proche gare et commerces. Cave et buanderie."
            ),
            "features": [
                "Vue Alpes",
                "Balcon",
                "Cuisine equipee",
                "Cave",
                "Animaux acceptes",
            ],
            "area": 78,
            "rooms": 3.5,
            "bathrooms": 1,
            "terrace": 8,
        },
        "agent_config": {
            "agent_name": "Sophie",
            "agency_name": "Avendo Immobilier",
            "transfer_number": "+41791234568",
            "notification_email": "team@avendo.ch",
            "hangup_instruction": "Merci pour votre interet. Nous vous recontacterons par email dans les 24 heures. Excellente journee!",
            "qualification_criteria": [
                {
                    "question": "Quel est votre budget mensuel maximum pour le loyer?",
                    "type": "number",
                    "expectedValue": "2500",
                    "eliminatory": True,
                    "rejectionMessage": "Malheureusement, le budget minimum requis pour ce bien est de 2500 CHF.",
                },
                {
                    "question": "Avez-vous un emploi stable ou un revenu regulier?",
                    "type": "yes_no",
                    "expectedValue": "oui",
                    "eliminatory": True,
                    "rejectionMessage": "Un emploi stable est requis pour ce bien.",
                },
                {
                    "question": "Combien de personnes occuperont le logement?",
                    "type": "number",
                    "expectedValue": "1-3",
                    "eliminatory": False,
                },
            ],
            "transfer_conditions": [
                "Lead wants to negotiate the rent",
                "Lead asks about specific renovation plans",
            ],
        },
        "viewing_slots": [
            {
                "date": tomorrow,
                "startTime": "14:00",
                "endTime": "15:00",
                "maxBookings": 3,
            },
            {
                "date": day_after,
                "startTime": "10:00",
                "endTime": "11:00",
                "maxBookings": 3,
            },
        ],
        "callback_webhook": "https://your-server.com/webhooks/nextimmolab",
        "external_ref": f"SAMPLE-{datetime.now().strftime('%Y%m%d-%H%M%S')}",
    }

    print(f"\nTriggering call to {payload['lead']['prenom']} {payload['lead']['nom']}...")
    res = requests.post(f"{API_BASE}/calls/trigger", headers=HEADERS, json=payload)

    if res.status_code != 200:
        print(f"Error {res.status_code}: {res.text}")
        sys.exit(1)

    data = res.json()
    print(f"Call triggered — ID: {data['callId']} | Status: {data['status']}")
    return data["callId"]


# ─── 3. Poll for results ────────────────────────────────────────────────────

def wait_for_result(call_id: str, timeout_min: int = 10):
    """GET /api/v1/calls/{callId} — poll until completed/failed."""

    print(f"\nPolling call {call_id[:8]}...")
    deadline = time.time() + timeout_min * 60

    while time.time() < deadline:
        res = requests.get(f"{API_BASE}/calls/{call_id}", headers=HEADERS)
        res.raise_for_status()
        call = res.json()

        status = call["status"]
        print(f"  [{datetime.now().strftime('%H:%M:%S')}] {status}", end="")

        if status in ("completed", "failed", "no_answer"):
            print()
            return call
        else:
            print(" — waiting 10s...")
            time.sleep(10)

    print("Timeout reached.")
    return None


# ─── 4. Display results ─────────────────────────────────────────────────────

def print_results(call: dict):
    """Pretty-print call results."""

    print("\n" + "=" * 60)
    print(f"RESULT: {call['status'].upper()}")
    print("=" * 60)
    print(f"Lead:      {call.get('leadName')}")
    print(f"Duration:  {call.get('duration', '--')}")
    print(f"Score:     {call.get('score', '--')}/100")
    print(f"Qualified: {call.get('qualified', '--')}")

    criteria = call.get("criteriaResults", [])
    if criteria:
        print(f"\nCriteria ({sum(1 for c in criteria if c['passed'])}/{len(criteria)} passed):")
        for c in criteria:
            icon = "PASS" if c["passed"] else "FAIL"
            elim = " [ELIMINATORY]" if c.get("eliminatory") else ""
            print(f"  [{icon}] {c['question']}{elim}")
            print(f"         Answer: {c.get('answer', '--')}")

    transcript = call.get("transcript", [])
    if transcript:
        print(f"\nTranscript ({len(transcript)} messages):")
        for t in transcript[:6]:
            role = "AGENT" if t["role"] == "agent" else "LEAD "
            print(f"  {role}: {t['text'][:80]}{'...' if len(t['text']) > 80 else ''}")
        if len(transcript) > 6:
            print(f"  ... +{len(transcript) - 6} more messages")

    print("=" * 60)


# ─── 5. List recent calls ───────────────────────────────────────────────────

def list_calls(status: str = None, limit: int = 5):
    """GET /api/v1/calls — list calls with optional filter."""

    params = {"limit": limit}
    if status:
        params["status"] = status

    res = requests.get(f"{API_BASE}/calls", headers=HEADERS, params=params)
    res.raise_for_status()
    data = res.json()

    print(f"\nRecent calls ({data['total']} total):")
    for c in data["calls"]:
        score = f"score={c.get('score', '--')}" if c["status"] == "completed" else ""
        print(f"  {c['id'][:8]}... | {c['status']:12} | {c['leadName']:20} | {score}")


# ─── Main ────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    check_usage()
    call_id = trigger_call()

    print("\nWaiting 5s for call to connect...")
    time.sleep(5)

    result = wait_for_result(call_id)
    if result:
        print_results(result)

    list_calls(limit=5)
